### PR TITLE
[Flare] Update interactiveUpdates flushing heuristics

### PR DESCRIPTION
**Pull request by trueadm**
This PR changes the logic that occurs in `interactiveUpdates`, to account for a scenario that proved problematic in the testing of React Flare – the mixing of both event systems on the same app. Note: this change is behind the experimental event API. Also included in this PR are:

- `interactiveUpdates` -> `discreteUpdates`
- `flushInteractiveUpdates` -> `flushDiscreteUpdates`
- Adds `batchedEventUpdates` and a new `BatchedEventPhase` to scheduler

## Discrete events use timeStamp to manage flushing

Now that the Flare event system and the existing event system co-exists, we've noticed far more flushing is occurring; which is problematic for cases where this behaviour was non-existant with only the existing event system. To counter this a `timeStamp` heuristic has been added by reading the `timeStamp` from the `window` object. We will probably want to change this to be an argument (rather than reading from window) in the future, but I didn't want to introduce too many changes in this PR.

The `timeStamp` is an important field on events, as it allows us to tell what events were part of the same real-world interaction. For example: `mouseup`, `pointerup` and `click` all should have the same `timeStamp`. If we know that they're all the same `timeStamp`, we can do a single flush for them all, which improves compatibility with Flare and also should improve runtime performance as we're flushing sync less.

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 1]
```diff
@@ -604,20 +605,33 @@ export function deferredUpdates<A>(fn: () => A): A {
...
   fn: (A, B, C) => R,
   a: A,
   b: B,
   c: C,
 ): R {
-  if (workPhase === NotWorking) {
-    // TODO: Remove this call. Instead of doing this automatically, the caller
-    // should explicitly call flushInteractiveUpdates.
-    flushPendingDiscreteUpdates();
+  const currentEvent = typeof window !== 'undefined' && window.event;
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: Let's lift this logic into React DOM/Fire. The renderer can export two methods: `interactiveUpdates` and `flushInteractiveUpdates`.

Also maybe we should go with "discrete"? We've been using both but I kinda like "discrete" more. Not required to decide on that in this PR, though.

File: packages/react-events/src/__tests__/Press-test.internal.js [hunk 2]
```diff
@@ -2301,4 +2303,58 @@ describe('Event responder: Press', () => {
...
+    dispatchEvent('pointerdown');
+    dispatchEvent('pointerup');
+    dispatchEvent('click');
+
+    expect(renderCounts).toBe(2);
+    Scheduler.flushAll();
+    expect(renderCounts).toBe(4);
+
+    document.body.removeChild(newContainer);
+  });
```
Comment (acdlite) [inline packages/react-events/src/__tests__/Press-test.internal.js]: Can you add a test that includes a subsequent event sequence? To check that `interactiveEventTimeStamp` gets reset properly.

**Comment by acdlite** [comment]
Also how confident are we that this works in non-Chrome browsers?

File: packages/react-events/src/__tests__/Press-test.internal.js (hunk 2 above)
Comment (trueadm) [reply packages/react-events/src/__tests__/Press-test.internal.js]: Added some further tests to verify this works.

File: packages/react-reconciler/src/ReactFiberScheduler.js (hunk 1 above)
Comment (trueadm) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: Done :)

**Comment by trueadm** [comment]
@acdlite I checked it Safari, FF, Chrome and Edge. All looking good with this change. I need to confirm with IE, but this is behind a flag anyway, so there should be no immediate risk.

File: packages/react-events/src/__tests__/Press-test.internal.js [hunk 3]
```diff
@@ -2347,13 +2345,32 @@ describe('Event responder: Press', () => {
...
-    expect(renderCounts).toBe(4);
+    if (__DEV__) {
+      expect(renderCounts).toBe(4);
+    } else {
+      expect(renderCounts).toBe(2);
+    }
+
+    dispatchEvent('pointerdown', 100);
+    dispatchEvent('pointerup', 100);
+    // Ensure the timeStamp logic works
```
Comment (necolas) [inline packages/react-events/src/__tests__/Press-test.internal.js]: Is the Press event component the best place to test this logic?

Comment (trueadm) [reply packages/react-events/src/__tests__/Press-test.internal.js]: It’s fine for now. All responders should have some variant of this too.

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 4]
```diff
@@ -566,8 +566,8 @@ export function flushRoot(root: FiberRoot, expirationTime: ExpirationTime) {
   flushSyncCallbackQueue();
 }
 
-export function flushInteractiveUpdates() {
-  if (workPhase === RenderPhase || workPhase === CommitPhase) {
+export function flushDiscreteUpdates() {
+  if (workPhase !== NotWorking) {
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: Why did you change this? It should flush inside `unstable_batchedUpdates`, too.

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 5]
```diff
@@ -604,17 +604,12 @@ export function deferredUpdates<A>(fn: () => A): A {
...
   b: B,
   c: C,
 ): R {
-  if (workPhase === NotWorking) {
-    // TODO: Remove this call. Instead of doing this automatically, the caller
-    // should explicitly call flushInteractiveUpdates.
-    flushPendingDiscreteUpdates();
-  }
   if (!revertPassiveEffectsChange) {
     // TODO: Remove this call for the same reason as above.
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: Let's remove this `flushPassiveEffects` call now.

File: packages/events/ReactGenericBatching.js [hunk 6]
```diff
@@ -46,26 +47,31 @@ export function batchedUpdates(fn, bookkeeping) {
...
 
-export function interactiveUpdates(fn, a, b, c) {
-  return _interactiveUpdatesImpl(fn, a, b, c);
+export function discreteUpdates(fn, a, b, c) {
+  return discreteUpdatesImpl(fn, a, b, c);
 }
 
-export function flushInteractiveUpdates() {
-  return _flushInteractiveUpdatesImpl();
+let lastInteractiveTimeStamp = 0;
```
Comment (acdlite) [inline packages/events/ReactGenericBatching.js]: Hmm is this the right place to put this? (cc: @sebmarkbage) Seems good enough to merge but it wouldn't have been my first guess.

File: packages/react-reconciler/src/ReactFiberScheduler.js (hunk 4 above)
Comment (trueadm) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: In the previous `interactiveUpdates` function, we only called `flushInteractiveUpdates` when `workPhase` was not `NotWorking`. So this was to duplicate that behaviour or some `act` tests fails. Would you prefer we had another version of this function that is only called on the event path and thus revert `flushDiscreteUpdates` back to the old conditional?

File: packages/react-reconciler/src/ReactFiberScheduler.js (hunk 5 above)
Comment (trueadm) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: Remove `flushPassiveEffects` altogether and the flag, as it won't be used anywhere?

File: packages/events/ReactGenericBatching.js (hunk 6 above)
Comment (trueadm) [reply packages/events/ReactGenericBatching.js]: Seemed like the most sane place to put it, but I'm not bothered, it just meant less indirection on the existing codebase.

File: packages/react-reconciler/src/ReactFiberScheduler.js (hunk 4 above)
Comment (trueadm) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: Actually, I think I know what you meant now. Changes are reflected in latest commit.

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: The purpose of the check is to handle nested event dispatches, like `element.click()` or `element.focus()`, that are invoked inside a React lifecycle. In that case `workPhase === CommitPhase`; we can't synchronously flush pending discrete updates in that case because we're already in the middle of another render. So we exit and do nothing. Arguably, we should defer the nested event dispatch to a microtask, but that would be a breaking change in legacy mode. That seems outside the scope of this PR, though.

However, there's no valid case for `workPhase` to be `RenderPhase`. Without the check, this will trigger an internal invariant, but we should have a user-facing error with a helpful message.

So this is what I would suggest:

```js
if (workPhase === CommitPhase) {
  // We're inside the commit phase, so we can't synchronously flush pending
  // work. This is probably a nested event dispatch triggered by a
  // lifecycle/effect, like `el.focus()`. Exit.
  return;
}

if (workPhase === RenderPhase) {
  invariant(
    false,
    'unstable_flushDiscreteUpdates: Cannot flush updates when React is '
    + 'already rendering.'
  );
}

// Proceed with flushing
```

File: packages/react-reconciler/src/ReactFiberScheduler.js (hunk 5 above)
Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: Yeah

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: It's not a breaking change because this is only observable via the event system or via `unstable_interactiveUpdates`. After this PR, the event system will call `flushDiscreteUpdates`, which itself calls `flushPassiveEffects` behind a flag. So the behavior will be the same.

File: packages/react-reconciler/src/ReactFiberScheduler.js (hunk 4 above)
Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: Oops didn't see your comment while I was typing :D

**Comment by acdlite** [comment]
Can you add a test for:

```js
ReactDOM.unstable_batchedUpdates(() => {
  // This should flush synchronously
  ReactDOM.unstable_flushDiscreteUpdates();
  setState();
});
```

Or replace `batchedUpdates` with a click handler:

```js
function onClick() {
  // This should flush synchronously
  ReactDOM.unstable_flushDiscreteUpdates();
  setState();
};
```

I believe your current implementation of `flushDiscreteUpdates` would fail that test. See https://github.com/facebook/react/pull/15687#discussion_r285792879 for suggested implementation.

**Comment by acdlite** [comment]
Might also want to add a test case that includes an imperative `el.click()` inside an effect.

**Comment by acdlite** [comment]
Per our chat thread, I'd also be OK with adding a TODO above the early return for `BatchPhase` to remind ourselves to come back to it.

File: packages/react-events/src/__tests__/Press-test.internal.js (hunk 3 above)
Comment (necolas) [reply packages/react-events/src/__tests__/Press-test.internal.js]: It's that the changes are being made to the core but the tests are here. And those tests have `__DEV__`-specific results

File: packages/events/ReactGenericBatching.js (hunk 6 above)
Comment (trueadm) [reply packages/events/ReactGenericBatching.js]: Addressed this in latest commit.

**Comment by trueadm** [comment]
I've added more tests, updated tests, added `BatchedEventPhase` and also did various other things to tidy up the code.

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 7]
```diff
@@ -566,14 +567,20 @@ export function flushRoot(root: FiberRoot, expirationTime: ExpirationTime) {
...
 }
 
-export function flushInteractiveUpdates() {
-  if (workPhase === RenderPhase || workPhase === CommitPhase) {
-    // Can't synchronously flush interactive updates if React is already
-    // working. This is currently a no-op.
-    // TODO: Should we fire a warning? This happens if you synchronously invoke
-    // an input event inside an effect, like with `element.click()`.
+export function flushDiscreteUpdates() {
+  if (workPhase === CommitPhase || workPhase === BatchedPhase) {
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: Why is this a no-op in `batchedUpdates`? Is there a test that failed otherwise?

Comment (trueadm) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: Yep, a bunch of tests failed from memory.

**Comment by tjallingt** [comment]
I've previously encountered some issues related to Event.timeStamp in Vue with forks of older Chrome versions (specifically QtWebEngine) where the timeStamp would be some very large **negative** number.

Might be interesting to take a look at these:
 
https://github.com/vuejs/vue/issues/9681
https://github.com/vuejs/vue/commit/7591b9dc6dde314f2d32dcd7a8355f696a330979
https://github.com/vuejs/vue/blob/dev/src/platforms/web/runtime/modules/events.js#L42-L90

**Comment by trueadm** [comment]
I saw some prior issues in Firefox, where the time stamps where using different precision timers (this longer numbers for some events than others). I believe this got fixed in Firefox after 52; there was a bug report I doh f earlier that stated the issue was resolved.

**Comment by trueadm** [comment]
I saw some prior issues in Firefox, where the time stamps where using different precision timers (this longer numbers for some events than others). I believe this got fixed in Firefox after 52; there was a bug report I doh f earlier that stated the issue was resolved.

File: packages/react-reconciler/src/ReactFiberScheduler.js (hunk 7 above)
Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: Can you put a TODO there? I'll take a look later. Doesn't have to block merge.

File: packages/react-events/src/__tests__/Press-test.internal.js (hunk 3 above)
Comment (trueadm) [reply packages/react-events/src/__tests__/Press-test.internal.js]: We can follow up after and make tests in core too. I don't want to block this any longer given I'm on PTO in a few days. Maybe we can add this to our TODO list otherwise.

File: packages/react-reconciler/src/ReactFiberScheduler.js (hunk 7 above)
Comment (trueadm) [reply packages/react-reconciler/src/ReactFiberScheduler.js]: Added a TODO. Let's resync with this once I'm back from PTO; this at least unblocks us internally.

File: packages/react-dom/src/events/DOMEventResponderSystem.js [hunk 8]
```diff
@@ -990,3 +998,29 @@ export function generateListeningKey(
   const passiveKey = passive ? '_passive' : '_active';
   return `${topLevelType}${passiveKey}`;
 }
+
+let lastDiscreteEventTimeStamp = 0;
+let useTimeStampHeuristic = true;
+
+if (canUseDOM) {
+  const firefoxUA = navigator.userAgent.match(/Firefox\/(.*)$/);
```
Comment (acdlite) [inline packages/react-dom/src/events/DOMEventResponderSystem.js]: What will the behavior be in Firefox then? Will it be broken, or merely less optimal?

Also perhaps there's a way to detect this that doesn't rely on UA sniffing. Like by dispatching a click event on a dummy node and checking the timestamps of the events.

Comment (trueadm) [reply packages/react-dom/src/events/DOMEventResponderSystem.js]: Actually, I just managed to test on FF 51 again, we don't need this anymore. The timestamps might be different resolutions, as pointed out with Vue, but we don't care for the relevance in terms of difference (> or <), we only care for the equality (===).

File: packages/react-dom/src/events/DOMEventResponderSystem.js [hunk 9]
```diff
@@ -1001,6 +1001,16 @@ export function generateListeningKey(
...
+  // event.timeStamp isn't overly reliable due to inconsistencies in
+  // how different browsers have historically provided the time stamp.
+  // Some browsers provide high-resolution time stamps for all events,
+  // some provide low-resoltion time stamps for all events. FF < 52
+  // even mixes both time stamps together. Some browsers even report
+  // negative time stamps or time stamps that are 0 (iOS9) in some cases.
+  // Given we are only comparing two time stamps with equality (!==),
+  // we are safe from the resolution differences. If the time stamp is 0
+  // we bail-out of preventing the flush, which shouldn't have any
+  // effect on the desired output, other than we over-flush.
```
Comment (acdlite) [inline packages/react-dom/src/events/DOMEventResponderSystem.js]: The last sentence implies that it's only a performance change, which isn't true. It can affect semantics as well, such as if an earlier flush removes or adds event listeners that are fired in the subsequent flush.

**Comment by acdlite** [comment]
^ We need to think about this heuristic more before we ship this to open source but since this is gated behind the Flare feature flag, it's good to merge so we can test it internally.
